File: Problems 1200-1299/1298. Maximum Candies You Can Get from Boxes/solution.cpp
class Solution {
    public:
        int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
            int n = status.size();
            vector<bool> have(n, false);
            queue<int> q;
    
            for (int box : initialBoxes) {
                if (status[box]) {
                    q.push(box);
                }
                have[box] = true;
            }
            int res = 0;
            while (!q.empty()) {
                int box = q.front();
                res += candies[box];
                q.pop();
    
                for (int key : keys[box]) {
                    if (have[key] && !status[key]) {
                        q.push(key);
                    }
                    status[key] = 1;
                }
    
                for (int contained : containedBoxes[box]) {
                    if (status[contained]) {
                        q.push(contained);
                    }
                    have[contained] = true;
                }
            }
            return res;
        }
    };

File: Problems 1200-1299/1298. Maximum Candies You Can Get from Boxes/solution.py
class Solution:
    def maxCandies(self, status: List[int], candies: List[int], keys: List[List[int]], containedBoxes: List[List[int]], initialBoxes: List[int]) -> int:
        n = len(status)
        have = [False] * n
        q = deque()
        
        for box in initialBoxes:
            # box is opened, add to queue
            if status[box] == 1:
                q.append(box)
            have[box] = True
        
        res = 0
        while q:
            box = q.popleft()
            res += candies[box]

            # Process any boxes you had keys for previously
            # but could not open because you didn't have the
            # box yet

            for key in keys[box]:
                # Can only open box if you have it
                # Check status to make sure we're not
                # processing same box twice (only process
                # unopened boxes)
                if have[key] and status[key] == 0:
                    q.append(key)
                # Set it as "opened", even if we haven't
                # found box yet, if we find it contained
                # in a box later on, it'll be added to the queue
                status[key] = 1
            
            for contained in containedBoxes[box]:
                if status[contained] == 1:
                    q.append(contained)
                have[contained] = True
        
        return res